dask_expr/_expr.py

```diff
@@ -3216,6 +3082,216 @@ def _kwargs(self) -> dict:
+        return calc_divisions_for_align(*self.args)
+
+    def _simplify_up(self, parent, dependents):
+        if isinstance(parent, Projection) and self._projection_passthrough:
+            return plain_column_projection(self, parent, dependents)
+
+    @functools.cached_property
+    def args(self):
+        dfs = [op for op in self.operands if isinstance(op, Expr)]
+        return [op for op in dfs if not is_broadcastable(dfs, op)]
+
+    def _lower(self):
+        # This can be expensive when something that has expensive division
+        # calculation is in the Expression
+        dfs = self.args
+        if (
+            len(dfs) == 1
+            or all(
+                dfs[0].divisions == df.divisions and df.known_divisions for df in dfs
+            )
+            or len(self.divisions) == 2
+        ):
+            return self._expr_cls(*self.operands)
+        args = maybe_align_partitions(*self.operands, divisions=self.divisions)
+        return self._expr_cls(*args)
+
+    @functools.cached_property
+    def _meta(self):
+        return self._expr_cls(*self.operands)._meta
+
+
+class CombineFirstAlign(MaybeAlignPartitions):
+    _parameters = ["frame", "other"]
+    _expr_cls = CombineFirst
+
+    def _simplify_up(self, parent, dependents):
+        # TODO: de-duplicate
+        if isinstance(parent, Projection):
+            columns = determine_column_projection(self, parent, dependents)
+            frame_columns = [col for col in self.frame.columns if col in columns]
+            other_columns = [col for col in self.other.columns if col in columns]
+            if (
+                self.frame.columns == frame_columns
+                and self.other.columns == other_columns
+            ):
+                return
+
+            return type(parent)(
+                type(self)(self.frame[frame_columns], self.other[other_columns]),
+                *parent.operands[1:],
+            )
+
+
+class FillnaAlign(MaybeAlignPartitions):
+    _projection_passthrough = True
+    _parameters = ["frame", "value"]
+    _expr_cls = Fillna
+
+
+class AlignAlignPartitions(MaybeAlignPartitions):
+    _parameters = ["frame", "other", "join", "axis", "fill_value"]
+    _expr_cls = _Align
+
+
+class CombineSeriesAlign(MaybeAlignPartitions):
+    _parameters = ["frame", "other", "func", "fill_value"]
+    _expr_cls = CombineSeries
+
+
+class CombineFrameAlign(MaybeAlignPartitions):
+    _parameters = ["frame", "other", "func", "fill_value", "overwrite"]
+    _expr_cls = CombineSeries
+
+
+class AssignAlign(MaybeAlignPartitions):
+    _parameters = ["frame", "column", "value"]
+    _expr_cls = Assign
+
+    def _simplify_up(self, parent, dependents):
+        # TODO: de-duplicate
+        if isinstance(parent, Projection):
+            columns = determine_column_projection(self, parent, dependents)
+            if not isinstance(columns, list):
+                columns = [columns]
+
+            cols = set(columns) - set(self.keys)
```

Comment:
This class has no `keys` property, so this PR causes a new failure in dask-cudf (I'll try to make a simple repro).